File: crates/remote-protocol/src/messages.rs
use crate::capabilities::Capability;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MessageEnvelope {
    pub v: u8,
    pub id: String,
    pub timestamp: u64,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub data: serde_json::Value,
}

// Client -> Server Data Payloads

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PointerDeltaData {
    pub dx: f32,
    pub dy: f32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub dt: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PointerButtonData {
    pub button: String, // "left", "right", "middle"
    pub state: String,  // "down", "up", "click", "double_click"
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PointerScrollData {
    pub dx: f32,
    pub dy: f32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct KeyActionData {
    pub key: String,
    pub state: String, // "down", "up", "tap"
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub modifiers: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextStreamData {
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MediaCommandData {
    pub action: String, // "play", "pause", "play_pause", "next", "prev", "volume_up", "volume_down", "set_volume", "mute", "seek"
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub value: Option<f32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PresentationCommandData {
    pub action: String, // "next", "prev", "start", "stop", "black_screen", "goto_slide"
    #[serde(rename = "slideIndex", default, skip_serializing_if = "Option::is_none")]
    pub slide_index: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WindowActionData {
    #[serde(rename = "windowId")]
    pub window_id: String,
    pub action: String, // "focus", "minimize", "maximize", "restore", "close", "snap_left", "snap_right", "move_to_display"
    #[serde(rename = "targetDisplay", default, skip_serializing_if = "Option::is_none")]
    pub target_display: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PairRequestData {
    #[serde(rename = "clientId")]
    pub client_id: String,
    #[serde(rename = "clientName")]
    pub client_name: String,
    pub token: String,
    #[serde(rename = "publicKey")]
    pub public_key: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LoginChallengeData {
    pub nonce: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LoginResponseData {
    #[serde(rename = "clientId")]
    pub client_id: String,
    pub signature: String,
    pub nonce: String,
}

// Server -> Client Data Payloads

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SessionReadyData {
    #[serde(rename = "serverName")]
    pub server_name: String,
    #[serde(rename = "serverVersion")]
    pub server_version: String,
    pub capabilities: Vec<Capability>,
    #[serde(rename = "activeDisplayCount")]
    pub active_display_count: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ForegroundAppState {
    #[serde(rename = "processName")]
    pub process_name: String,
    #[serde(rename = "windowTitle")]
    pub window_title: String,
    pub category: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MediaSessionState {
    pub title: String,
    pub artist: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub album: Option<String>,
    #[serde(rename = "isPlaying")]
    pub is_playing: bool,
    #[serde(rename = "positionSec")]
    pub position_sec: f32,
    #[serde(rename = "durationSec")]
    pub duration_sec: f32,
    pub volume: f32,
    #[serde(rename = "sourceApp", default, skip_serializing_if = "Option::is_none")]
    pub source_app: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DisplayInfo {
    pub index: u32,
    pub name: String,
    pub width: u32,
    pub height: u32,
    #[serde(rename = "isPrimary")]
    pub is_primary: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FileItem {
    pub id: String,
    pub name: String,
    #[serde(rename = "isDir")]
    pub is_dir: bool,
    #[serde(rename = "sizeBytes", default, skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<u64>,
    #[serde(rename = "modifiedAt", default, skip_serializing_if = "Option::is_none")]
    pub modified_at: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extension: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct VirtualRoot {
    pub id: String,
    pub name: String,
    #[serde(rename = "pathAlias")]
    pub path_alias: String,
}

File: packages/protocol/src/messages.ts
import { Capability } from './capabilities';

export type MessageType =
  // Realtime
  | 'input.pointer.delta'
  | 'input.pointer.button'
  | 'input.pointer.scroll'
  | 'keyboard.key'
  | 'keyboard.text'
  // Control
  | 'media.command'
  | 'presentation.command'
  | 'apps.launch'
  | 'windows.action'
  | 'macro.execute'
  | 'power.command'
  // Data
  | 'clipboard.get'
  | 'clipboard.set'
  | 'files.list_roots'
  | 'files.roots'
  | 'files.browse'
  | 'files.items'
  | 'files.read_chunk'
  | 'files.read_file'
  | 'files.content'
  | 'panels.list'
  | 'panels.save'
  // Auth & Session
  | 'auth.pair_request'
  | 'auth.pair_response'
  | 'auth.login_challenge'
  | 'auth.login_response'
  | 'auth.session_ready'
  | 'auth.error'
  // State Sync Broadcasts
  | 'state.foreground_app'
  | 'state.media_session'
  | 'state.displays'
  | 'state.clipboard'
  | 'state.system_status'
  // Generic Acknowledgement
  | 'ack'
  | 'error';

export interface MessageEnvelope<T = unknown> {
  v: 1;
  id: string;
  timestamp: number;
  type: MessageType;
  data: T;
}

// Client -> Server Data Payloads
export interface PointerDeltaData {
  dx: number;
  dy: number;
  dt?: number;
}

export interface PointerButtonData {
  button: 'left' | 'right' | 'middle';
  state: 'down' | 'up' | 'click' | 'double_click';
}

export interface PointerScrollData {
  dx: number;
  dy: number;
}

export interface KeyActionData {
  key: string;
  state: 'down' | 'up' | 'tap';
  modifiers?: Array<'ctrl' | 'alt' | 'shift' | 'win'>;
}

export interface TextStreamData {
  text: string;
}

export interface MediaCommandData {
  action:
    | 'play'
    | 'pause'
    | 'play_pause'
    | 'next'
    | 'prev'
    | 'volume_up'
    | 'volume_down'
    | 'set_volume'
    | 'mute'
    | 'seek';
  value?: number;
}

export interface PresentationCommandData {
  action: 'next' | 'prev' | 'start' | 'stop' | 'black_screen' | 'goto_slide';
  slideIndex?: number;
}

export interface WindowActionData {
  windowId: string;
  action:
    | 'focus'
    | 'minimize'
    | 'maximize'
    | 'restore'
    | 'close'
    | 'snap_left'
    | 'snap_right'
    | 'move_to_display';
  targetDisplay?: number;
}

export interface PairRequestData {
  clientId: string;
  clientName: string;
  token: string;
  publicKey: string;
}

export interface LoginChallengeData {
  nonce: string;
}

export interface LoginResponseData {
  clientId: string;
  signature: string;
  nonce: string;
}

// Server -> Client Data Payloads
export interface SessionReadyData {
  serverName: string;
  serverVersion: string;
  capabilities: Capability[];
  activeDisplayCount: number;
}

export interface ForegroundAppState {
  processName: string;
  windowTitle: string;
  category:
    | 'general'
    | 'media'
    | 'presentation'
    | 'browser'
    | 'development'
    | 'meeting'
    | 'design'
    | 'documents';
}

export interface MediaSessionState {
  title: string;
  artist: string;
  album?: string;
  isPlaying: boolean;
  positionSec: number;
  durationSec: number;
  volume: number;
  sourceApp?: string;
}

export interface DisplayInfo {
  index: number;
  name: string;
  width: number;
  height: number;
  isPrimary: boolean;
}

export interface FileItem {
  id: string;
  name: string;
  isDir: boolean;
  sizeBytes?: number;
  modifiedAt?: number;
  extension?: string;
}

export interface VirtualRoot {
  id: string;
  name: string;
  pathAlias: string;
}

export function createEnvelope<T>(type: MessageType, data: T): MessageEnvelope<T> {
  return {
    v: 1,
    id:
      typeof crypto !== 'undefined' && crypto.randomUUID
        ? crypto.randomUUID()
        : Math.random().toString(36).substring(2, 11),
    timestamp: Date.now(),
    type,
    data,
  };
}
